### PR TITLE
docs: application: document "make flash" conventions

### LINKED ISSUE #4652: Document "make flash" in the "application development primer"
"make flash" is an important feature of Zephyr, but the application development primer does not mention it.

Instead this feature is discoverable through various board and sample doc's.

make flash is so important that it should be mentioned in the application development primer.

### PR DESCRIPTION
**Pull request by mbolivar**
Add basic documentation for "make flash".

Fixes #4652.

### CODE REVIEW

File: doc/application/application.rst [hunk 1]
```diff
@@ -675,15 +675,48 @@ following procedure:
...
 ******************
 
-An application image can be run on real or emulated hardware. The kernel has
-built-in emulator support for QEMU. It allows you to run and test an application
-virtually, before (or in lieu of) loading and running it on actual target
-hardware.
+An application image can be run on real or emulated hardware.
+
+Most boards supported by Zephyr let you flash a compiled binary using
+the ``flash`` target to copy the binary to the board and run it.
```
Comment (dbkinder) [inline doc/application/application.rst]: to clarify, maybe better to say: 

      using the make ``flash`` target

File: doc/application/application.rst [hunk 2]
```diff
@@ -675,15 +675,48 @@ following procedure:
...
+Follow these instructions to flash and run an application on real
+hardware:
+
+#. Open a terminal console and navigate to the application directory
+   :file:`~/app`.
+
+#. Enter the following command to build and run the application on
+   your board, which you usually must attach via USB.
```
Comment (dbkinder) [inline doc/application/application.rst]: Either "usually" or "must", but not "usually must".  I'd pick "usually"

File: doc/application/application.rst [hunk 3]
```diff
@@ -675,15 +675,48 @@ following procedure:
...
-virtually, before (or in lieu of) loading and running it on actual target
-hardware.
+An application image can be run on real or emulated hardware.
+
+Most boards supported by Zephyr let you flash a compiled binary using
+the ``flash`` target to copy the binary to the board and run it.
+Follow these instructions to flash and run an application on real
+hardware:
+
+#. Open a terminal console and navigate to the application directory
```
Comment (dbkinder) [inline doc/application/application.rst]: To clarify, I'd suggest saying, 

     Open a terminal console on your host computer, and navigate ...

File: doc/application/application.rst [hunk 4]
```diff
@@ -675,15 +675,48 @@ following procedure:
...
+   board's documentation <boards>` for additional information on how
+   to flash your board.
+
+.. note:: When developing on Linux, it's common to need to install
+          board-specific udev rules to enable USB device access to
+          your board as a non-root user. If ``make flash`` fails,
+          consult your board's documentation to see if this is
+          necessary.
+
+The kernel has built-in emulator support for QEMU. It allows you to
```
Comment (dbkinder) [inline doc/application/application.rst]: Is it the kernel or the SDK that has this QEMU support?  (I think the latter).

**Review by dbkinder — DISMISSED** [review]
some suggested edits

File: doc/application/application.rst (hunk 1 above)
Comment (mbolivar) [reply doc/application/application.rst]: Yeah, I was trying to word this in a way that would be compatible with CMake (which supports Ninja in addition to Make), but this section will need to be revisited when that happens anyway.

File: doc/application/application.rst (hunk 3 above)
Comment (mbolivar) [reply doc/application/application.rst]: I'm just using the wording that was already in this document, but I agree that that's better. I'll make the same change below where it already appears for 'make run'.

File: doc/application/application.rst (hunk 2 above)
Comment (mbolivar) [reply doc/application/application.rst]: Whoops, nice catch; thanks.

File: doc/application/application.rst (hunk 4 above)
Comment (mbolivar) [reply doc/application/application.rst]: Note that I haven't altered this text; this sentence appeared as-is before I touched this file.

The kernel does have QEMU support, in the form of the qemu_xxx boards:

```
boards/arm/qemu_cortex_m3  boards/riscv32/qemu_riscv32  boards/xtensa/qemu_xtensa
boards/nios2/qemu_nios2    boards/x86/qemu_x86
```

If this sentence needs changing, could fixing it be handled separately from this PR? That's unrelated to documenting `make flash`

**Comment by mbolivar** [comment]
@dbkinder thanks for the review! I've made some changes and left comments regarding QEMU.

**Review by SebastianBoe — APPROVED** [review]
Great work!

**Comment by pfalcon** [comment]
Should there be more reviewers?

**Review by nashif — APPROVED** [review]

**Pull request merged**